osbenchmark/synthetic_data_generator/mapping_synthetic_data_generator.py

```diff
@@ -0,0 +1,438 @@
+    """
+    logger = logging.getLogger(__name__)
+
+    # Fetch settings and input config that user provided
+    generation_settings = get_generation_settings(input_config)
+    mapping_config = input_config
+
+    max_file_size_bytes = generation_settings.get('max_file_size_gb') * GB_TO_BYTES
+    total_size_bytes = sdg_config.total_size_gb * GB_TO_BYTES
+    docs_per_chunk = generation_settings.get('docs_per_chunk')
+    avg_document_size = get_avg_document_size(index_mappings, mapping_config)
+
+    current_size = 0
+    docs_written = 0
+    file_counter = 0
+    generated_dataset_details = []
+
+    logger.info("Average document size: %s", avg_document_size)
+    logger.info("Chunk size: %s docs", docs_per_chunk)
+    logger.info("Total GB to generate: %s", sdg_config.total_size_gb)
+    logger.info("Max file size in GB: %s", generation_settings.get('max_file_size_gb'))
+
+    console.println(f"Total GB to generate: {sdg_config.total_size_gb}\n"
+                    f"Average document size: {avg_document_size}\n"
+                    f"Max file size in GB: {generation_settings.get('max_file_size_gb')}\n")
+
+    start_time = time.time()
+    with tqdm(total=total_size_bytes,
+                unit='B',
+                unit_scale=True,
+                bar_format="{l_bar}{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}, {rate_fmt}]") as progress_bar:
+
+        setup_custom_tqdm_formatting(progress_bar)
+        while current_size < total_size_bytes:
+            file_path = os.path.join(sdg_config.output_path, f"{sdg_config.index_name}_{file_counter}.json")
+            file_size = 0
+            docs_written = 0
+
+            while file_size < max_file_size_bytes:
+                generation_start_time = time.time()
+                seeds = generate_seeds_for_workers(regenerate=True)
+                logger.info("Using seeds: %s", seeds)
+
+                # TODO: Submit the seeds to the client so that each client is producing different variations of documents
+                futures = [client.submit(MappingSyntheticDataGeneratorWorker.generate_documents_from_worker, index_mappings, mapping_config, docs_per_chunk) for seed in seeds]
+                results = client.gather(futures) # if using AS_COMPLETED remove this line
+
+                writing_start_time = time.time()
+                for data in results:
+                    written = write_chunk(data, file_path)
+                    docs_written += written
+                    written_size = written * avg_document_size
+                    current_size += written_size
+                    progress_bar.update(written_size)
+
+                writing_end_time = time.time()
+
+                file_size = os.path.getsize(file_path)
+                # If it exceeds the max file size, then append this to keep track of record
+                if file_size >= max_file_size_bytes:
+                    file_name = file_path.split("/")[-1]
```

Comment:
this will break on windows, better use `os.path.basename(file_path)`. Update everywhere.

Reply:
Wasn't aware but thanks for calling this out!